### ISSUE TITLE
Allow ledger to be updated in a safe way by state transfer

### ISSUE THREAD

**Issue by jyellick** [hyperledger-archives/fabric#1608]
## Description

When state transfer must perform a complete state snapshot transfer, it's not possible to do this in a safe way.

In particular, state snapshots come across the wire as a series of state deltas, and only one state delta can be applied before committing.  It is not possible to verify the state until the entire snapshot has been retrieved.  Therefore, we must commit unverified, certainly inconsistent, and possible wrong data to the ledger.

What is needed, is a way for state transfer to build up a new copy of the entire state that it can verify, and then atomically swap out for the current state.  In this way, queries against the 'stale' state can continue, and then once verified, the whole state can be made current.

The PR for #1557 causes queries to return error during state transfer because of the risk of returning inconsistent or wrong data.  This is overzealous, and this ledger enhancement would allow us to handle this state more elegantly.

I believe that this requires the expertise of @srderson

**Comment by christo4ferris** [comment]
@jyellick @srderson I presume that this is still relevant?

**Comment by jyellick** [comment]
Yes, and will likely be relevant in the next arch, though I will need to see a finalized pruning strategy before I can be certain.
